core/adminservice/server.go

```diff
@@ -0,0 +1,171 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package adminservice implements a set of functions that support backend integrations.
+package adminservice
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/golang/glog"
+	"github.com/google/keytransparency/core/crypto/commitments"
+	"github.com/google/keytransparency/core/crypto/signatures"
+	"github.com/google/keytransparency/core/crypto/vrf"
+	"github.com/google/keytransparency/core/mutator"
+	"github.com/google/keytransparency/core/mutator/entry"
+	tpb "github.com/google/keytransparency/core/proto/keytransparency_v1_types"
+	"github.com/google/keytransparency/core/transaction"
+	"github.com/google/trillian"
+)
+
+const (
+	maxBatchSize = 100
+)
+
+// Server implements an admin server for Key Tranpsarency.
+type Server struct {
+	mapID     int64
+	tmap      trillian.TrillianMapClient
+	logID     int64
+	tlog      trillian.TrillianLogClient
+	mutator   mutator.Mutator
+	mutations mutator.Mutation
+	factory   transaction.Factory
+
+	vrfPriv   vrf.PrivateKey
+	signers   []signatures.Signer
+	committer commitments.Committer
+}
+
+// New returns a new admin server.
+func New(mapID int64,
+	tmap trillian.TrillianMapClient,
+	logID int64,
+	tlog trillian.TrillianLogClient,
+	vrfPriv vrf.PrivateKey,
+	mutator mutator.Mutator,
+	mutations mutator.Mutation,
+	factory transaction.Factory,
+	committer commitments.Committer,
+	signers []signatures.Signer,
+) *Server {
+	return &Server{
+		mapID:     mapID,
+		tmap:      tmap,
+		logID:     logID,
+		tlog:      tlog,
+		vrfPriv:   vrfPriv,
+		mutator:   mutator,
+		mutations: mutations,
+		factory:   factory,
+		committer: committer,
+		signers:   signers,
+	}
+}
+
+// BatchUpdateEntries performs batch updates to accounts using an authorized key.
+func (s *Server) BatchUpdateEntries(ctx context.Context,
+	in *tpb.BatchUpdateEntriesRequest) (*tpb.BatchUpdateEntriesResponse, error) {
+	// TODO(cessar): Authorize the incoming request via TLS client auth based on the appID.
+	if got, want := len(in.GetUsers()), maxBatchSize; got > want {
+		return nil, status.Errorf(codes.InvalidArgument, "len(users): %v, want < %v", got, want)
+	}
+	appID := in.GetAppId()
+	// Create UserID <-> index mapping
+	indexes := make([][]byte, 0, len(in.GetUsers()))
+	indexToUserID := make(map[string]string)
+	profiles := make(map[string]*tpb.UserProfile)
+	for userID, p := range in.GetUsers() {
+		index, _ := s.vrfPriv.Evaluate(vrf.UniqueID(userID, appID))
+		indexes = append(indexes, index[:])
+		indexToUserID[string(index[:])] = userID
+		profiles[userID] = p
+	}
+
+	getResp, err := s.tmap.GetLeaves(ctx, &trillian.GetMapLeavesRequest{
+		MapId:    s.mapID,
+		Index:    indexes,
+		Revision: -1,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create Update Requests.
```

Comment:
Please move what's inside the look to its own unexported function.

Reply:
done